### DISCUSSION TITLE
How to take backup and export data to other db

### DISCUSSION THREAD

**Discussion by apptonet** [pocketbase/pocketbase#115] [Q&A]
How to take backup and export data to other db?
how to set security set such as ip whitelist and black list

**Comment by ganigeorgiev** [comment]
Creating a backup with everything, including the the storage files, could be done by just copying the `pb_data` to your backup location.
Restore is the opposite (note the application shouldn't be running since it's not transactionally safe).

PocketBase internally uses SQLite and you can interact directly with the database (usually located at `pb_data/data.db`) via the `sqlite3` cli tool. For example here is using the `.backup` command:
```sh
sqlite3 /path/to/pb_data/data.db '.backup /path/to/backup'
```

sqlite3 cli has also `.dump` and `.output` commands but the result sql is not directly importable (it is just the data, aka. there are no DELETE TABLE -> CREATE TABLE -> INSERT ROWS sequence as in `mysqldump` for example), so you'll have to read and slightly adjust the file to import it in other databases.

You can also use some smarter DB management platforms like DBeaver that has its own importing/export tools.

---

For the IP whitelist/blacklist - this usually could be done by either by using a reverse proxy (nginx, apache, etc.), OR use PocketBase as a framework and attach a [echo](https://github.com/labstack/echo) middleware in the `OnBeforeServe` hook, eg:
```go
package main

import (
    "log"
    "net/http"

    "github.com/labstack/echo/v5"
    "github.com/pocketbase/pocketbase"
    "github.com/pocketbase/pocketbase/apis"
    "github.com/pocketbase/pocketbase/core"
)

func ipCheck(app core.App) echo.MiddlewareFunc {
    return func(next echo.HandlerFunc) echo.HandlerFunc {
        return func(c echo.Context) error {
            // do something with the request data...

            return next(c)
        }
    }
}

func main() {
    app := pocketbase.New()

    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        e.Router.Pre(ipCheck(app))

        return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```
